### ISSUE TITLE
Inconsistency between Timestamp and SystemTime use of seconds/nanos fields

### ISSUE THREAD

**Issue by joseph-wakeling-frequenz** [danburkert/prost#327]
This behaviour was observed with prost_types 0.6.1 on POSIX (Ubuntu 18.04).  The behaviour may have changed since https://github.com/danburkert/prost/pull/297 was merged but I'm reporting it on the assumption that it's likely a separate issue.

I was interested in the behaviour of `TryFrom<Timestamp> for SystemTime` when `Timestamp.seconds` was negative, so tried this test case:

```Rust
let ts = Timestamp { seconds: -20, nanos: 999 };
let systime = match SystemTime::try_from(timestamp) {
    Ok(systime) => systime,
    Err(duration) => UNIX_EPOCH - duration
}
```

This was motivated by the documented behaviour that "if the timestamp falls before the Unix epoch, [`try_from` returns] a duration containing the difference".  So it would make sense that one could just take `UNIX_EPOCH` and subtract the duration from it, to get the matching `SystemTime`.

However, the result differed from the `Timestamp` info in an interesting way:

```
SystemTime { tv_sec: -21, tv_nsec: 999999001 }
```

In other words, the `SystemTime` appears to treat its seconds and nanos parameters as separate additions to the Unix epoch ("subtract 21 seconds from the Unix epoch and add 999_999_001 nanos") while `TimeStamp` appears to treat its `nanos` as effectively decimal places for its `seconds` field, implicitly sharing its sign ("subtract 20 seconds and 999 nanos from Unix epoch").

This would appear to violate the [protobuf spec](https://developers.google.com/protocol-buffers/docs/reference/csharp/class/google/protobuf/well-known-types/timestamp#nanos), which reads:

> Negative second values with fractions must still have non-negative nanos values that count ***forward in time***. Must be from 0 to 999,999,999 inclusive. 

(my emphasis).  In other words, `Timestamp { seconds: -20, nanos: 999 }` ought to mean "subtract 20 seconds from the Unix epoch and add 999 nanos", and hence conversion to `SystemTime` would result in a matching `SystemTime { tv_sec: -20, tv_nsec: 999 }`.

I haven't tested yet with prost_types master, so it may also be that this discrepancy has vanished as part of the changes in https://github.com/danburkert/prost/pull/297.  But the existing test case:
https://github.com/danburkert/prost/blob/80256277997975948d257faf3f35c2890bf12787/prost-types/src/lib.rs#L160-L177

... won't catch this discrepancy, because it merely validates that converting back and forth gives the same result one started with, not that the actual seconds and nanos fields match on either side of the conversion.

**Comment by danburkert** [comment]
I think this is fixed in #328.

**Comment by joseph-wakeling-frequenz** [comment]
@danburkert it was actually already fixed by you in #297: my PR #328 just added test cases to make sure it didn't regress in future.  You gave me a little more credit than I deserved in the 0.7.0 release notes (but thank you in any case for the acknowledgement:-)

Thanks in any case for the new release, and Happy New Year to you and yours!
